extensions/librdkafka/PublishKafka.cpp

```diff
@@ -289,122 +356,309 @@ bool PublishKafka::configureNewConnection(const std::shared_ptr<KafkaConnection>
-          logger_->log_error("PublishKafka: configure delivery guarantee error result [%s]", errstr);
-      }
-      value = "";
-      if (context->getProperty(RequestTimeOut, value, flowFile) && !value.empty()) {
-        core::TimeUnit unit;
-        if (core::Property::StringToTime(value, valInt, unit) && core::Property::ConvertTimeUnitToMS(valInt, unit, valInt)) {
-          valueConf = std::to_string(valInt);
-          rd_kafka_topic_conf_set(topic_conf_, "request.timeout.ms", valueConf.c_str(), errstr, sizeof(errstr));
-          logger_->log_debug("PublishKafka: request.timeout.ms [%s]", valueConf);
-          if (result != RD_KAFKA_CONF_OK)
-            logger_->log_error("PublishKafka: configure timeout error result [%s]", errstr);
-        }
-      }
+  // Get some properties not (only) used directly to set up librdkafka
+  std::string value;
+
+  // Batch Size
+  uint32_t batch_size;
+  value = "";
+  if (context->getProperty(BatchSize.getName(), value) && !value.empty() && core::Property::StringToInt(value, batch_size)) {
+    logger_->log_debug("PublishKafka: Batch Size [%lu]", batch_size);
+  } else {
+    batch_size = 10;
+  }
 
-      std::shared_ptr<KafkaTopic> kafkaTopicref = std::make_shared<KafkaTopic>(topic_reference, topic_conf_);
+  // Target Batch Payload Size
+  uint64_t target_batch_payload_size;
+  value = "";
+  if (context->getProperty(TargetBatchPayloadSize.getName(), value) && !value.empty() && core::Property::StringToInt(value, target_batch_payload_size)) {
+    logger_->log_debug("PublishKafka: Target Batch Payload Size [%llu]", target_batch_payload_size);
+  } else {
+    target_batch_payload_size = 512 * 1024U;
+  }
 
-      conn->putTopic(topic, kafkaTopicref);
-    }
+  // Max Flow Segment Size
+  uint64_t max_flow_seg_size;
+  value = "";
+  if (context->getProperty(MaxFlowSegSize.getName(), value) && !value.empty() && core::Property::StringToInt(value, max_flow_seg_size)) {
+    logger_->log_debug("PublishKafka: Max Flow Segment Size [%llu]", max_flow_seg_size);
   } else {
-    logger_->log_error("Do not have required properties");
-    session->transfer(flowFile, Failure);
+    max_flow_seg_size = 0U;
+  }
+
+  // Attributes to Send as Headers
+  utils::Regex attributeNameRegex;
+  value = "";
+  if (context->getProperty(AttributeNameRegex.getName(), value) && !value.empty()) {
+    attributeNameRegex = utils::Regex(value);
+    logger_->log_debug("PublishKafka: AttributeNameRegex [%s]", value);
+  }
+
+  // Collect FlowFiles to process
+  uint64_t actual_bytes = 0U;
+  std::vector<std::shared_ptr<core::FlowFile>> flowFiles;
+  for (uint32_t i = 0; i < batch_size; i++) {
+    std::shared_ptr<core::FlowFile> flowFile = session->get();
+    if (flowFile == nullptr) {
+      break;
+    }
+    actual_bytes += flowFile->getSize();
+    flowFiles.emplace_back(std::move(flowFile));
+    if (target_batch_payload_size != 0U && actual_bytes >= target_batch_payload_size) {
```

Comment:
As far as I understand we can exceed the batch size here. 
I don't think there is a session API call to provide the size of the next flowfile without removing that from session. 
I think we should address it in a follow-up as it's ugly to exceed the configured limit, although it's not that urgent.

Reply:
Agreed, but that's why I sneakily called the property *Target* Batch Payload Size and not *Max* Batch Payload Size.